modules/geoip2/maxminddb-helper.c

```diff
@@ -0,0 +1,355 @@
+    }
+    break;
+    case MMDB_DATA_TYPE_BYTES:
+    case MMDB_DATA_TYPE_BOOLEAN:
+    case MMDB_DATA_TYPE_UINT128:
+    case MMDB_DATA_TYPE_UTF8_STRING:
+    case MMDB_DATA_TYPE_DOUBLE:
+    case MMDB_DATA_TYPE_FLOAT:
+    case MMDB_DATA_TYPE_UINT16:
+    case MMDB_DATA_TYPE_UINT32:
+    case MMDB_DATA_TYPE_UINT64:
+    case MMDB_DATA_TYPE_INT32:
+      entry_data_list = entry_data_list->next;
+      break;
+
+    default:
+      *status = MMDB_INVALID_DATA_ERROR;
+      return NULL;
+    }
+
+  *status = MMDB_SUCCESS;
+  return entry_data_list;
+}
+
+static void
+_geoip_log_msg_add_value(LogMessage *msg, GArray *path, GString *value)
+{
+  gchar *path_string = g_strjoinv(".", (gchar **)path->data);
+  log_msg_set_value_by_name(msg, path_string, value->str, value->len);
+  g_free(path_string);
+}
+
+static MMDB_entry_data_list_s *
+select_language(LogMessage *msg, MMDB_entry_data_list_s *entry_data_list, GArray *path, gint *status)
+{
+
+  if (entry_data_list->entry_data.type != MMDB_DATA_TYPE_MAP)
+    {
+      *status = MMDB_INVALID_DATA_ERROR;
+      return NULL;
+    }
+
+  guint32 size = entry_data_list->entry_data.data_size;
+
+  for (entry_data_list = entry_data_list->next;
+       size && entry_data_list; size--)
+    {
+
+      if (entry_data_list->entry_data.type != MMDB_DATA_TYPE_UTF8_STRING)
+        {
+          *status = MMDB_INVALID_DATA_ERROR;
+          return NULL;
+        }
+
+      GString *key = scratch_buffers_alloc();
+      g_string_printf(key, "%.*s",
+                      entry_data_list->entry_data.data_size,
+                      entry_data_list->entry_data.utf8_string);
+
+      entry_data_list = entry_data_list->next;
+      gchar *preferred_language = "en";
+      if (!strcmp(key->str, preferred_language))
+        {
+          if (entry_data_list->entry_data.type != MMDB_DATA_TYPE_UTF8_STRING)
+            {
+              *status = MMDB_INVALID_DATA_ERROR;
+              return NULL;
+            }
+
+          g_array_append_val(path, preferred_language);
+
+          GString *value = scratch_buffers_alloc();
+          g_string_printf(value, "%.*s",
+                          entry_data_list->entry_data.data_size,
+                          entry_data_list->entry_data.utf8_string);
+          _geoip_log_msg_add_value(msg, path, value);
+          g_array_remove_index(path, path->len-1);
+
+          entry_data_list = entry_data_list->next;
+        }
+      else
+        {
+          entry_data_list = mmdb_skip_tree(entry_data_list, status);
+          if (MMDB_SUCCESS != *status)
+            return NULL;
+        }
+    }
+  return entry_data_list;
+}
+
+static void
+_index_array_in_path(GArray *path, guint32 _index, GString *indexer)
+{
+  g_string_printf(indexer, "%d", _index);
+  ((gchar **)path->data)[path->len-1] = indexer->str;
+}
+
+MMDB_entry_data_list_s *
+dump_geodata_into_msg(LogMessage *msg, MMDB_entry_data_list_s *entry_data_list, GArray *path, gint *status)
+{
+  GString *value = NULL;
```

Comment:
This function is ... huge... :)
And there are copy-paste-like case branches (I know that the problem is that the TYPE and the format string and the name of the union member are not mapped).